File: java/src/main/java/dev/sanmer/geomag/Geomag.java
package dev.sanmer.geomag;

import java.time.LocalDateTime;

public class Geomag {
    static {
        Library.load();
    }

    public static native double toDecimalYears(int year, int month, int day, int hour, int minute, int second);
    public static native MagneticField wmm(double longitude, double latitude, double altitude, double decimalYears);
    public static native MagneticField igrf(double longitude, double latitude, double altitude, double decimalYears);

    public static double toDecimalYears(LocalDateTime date) {
        return toDecimalYears(date.getYear(), date.getMonthValue(), date.getDayOfMonth(), date.getHour(), date.getMinute(), date.getSecond());
    }

    public static MagneticField wmm(double longitude, double latitude, double altitude, LocalDateTime date) {
        return wmm(longitude, latitude, altitude, toDecimalYears(date));
    }

    public static MagneticField igrf(double longitude, double latitude, double altitude, LocalDateTime date) {
        return igrf(longitude, latitude, altitude, toDecimalYears(date));
    }
}

File: src/commonMain/kotlin/dev/sanmer/geomag/Geomag.kt
package dev.sanmer.geomag

import kotlinx.datetime.LocalDateTime


object Geomag {
    init {
        ImplLibrary.load()
    }

    external fun toDecimalYears(year: Int, month: Int, day: Int, hour: Int, minute: Int, second: Int): Double
    external fun wmm(longitude: Double, latitude: Double, altitude: Double, decimalYears: Double): MagneticField
    external fun igrf(longitude: Double, latitude: Double, altitude: Double, decimalYears: Double): MagneticField

    fun toDecimalYears(date: LocalDateTime) =
        toDecimalYears(date.year, date.monthNumber, date.dayOfMonth, date.hour, date.minute, date.second)

    fun wmm(longitude: Double, latitude: Double, altitude: Double, date: LocalDateTime) =
        wmm(longitude, latitude, altitude, toDecimalYears(date))

    fun igrf(longitude: Double, latitude: Double, altitude: Double, date: LocalDateTime) =
        igrf(longitude, latitude, altitude, toDecimalYears(date))
}